New example for gaussian_filter

kolibril13 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
https://github.com/scipy/scipy/issues/15032 , @tupui

#### What does this implement/fix?
<!--Please explain your changes.-->
Additional plot for explaining smoothing filter in 3d with `gaussian_filter` 
<!--Any additional information you think is important.-->
![image](https://user-images.githubusercontent.com/44469195/141698380-1c7f8cf1-97f3-4df2-a0ce-2439148a0f91.png)

tylerjereddy commented:
Looks like the CI complaints (so far) are mostly superficial linting issues so may make sense to just wait to see if there's any more meaningful feedback before fixing that up.

tylerjereddy commented on doc/source/tutorial/examples/gaussian_filter_plot1.py:
```diff
@@ -0,0 +1,62 @@
+voxelarray[cube((0, 0, boxs))] = 5 * 255 / 8
+voxelarray[cube((boxs, 0, boxs))] = 6 * 255 / 8
+voxelarray[cube((0, boxs, boxs))] = 7 * 255 / 8
+voxelarray[cube((boxs, boxs, boxs))] = 8 * 255 / 8
```
I guess the idea is to convey an instructional/educational pattern in the source here instead of using the simplest/"fastest" expression like `255` instead of `8 * 255 / 8` or `255 / 2` instead of `4 * 255 / 8`?

Or maybe it just doesn't matter much. I think Pamphile was already pretty happy with the final plot in any case, if I remember correctly. Maybe we'll just need to take a peek at the final rendered tutorial (the CI should link to an artifact/web page render when passing).

tylerjereddy commented:
This branch vs. `master` looks like an improvement:

![image](https://user-images.githubusercontent.com/7903078/141700013-bfda0012-c369-468b-8a06-2434943e14b5.png)

![image](https://user-images.githubusercontent.com/7903078/141700035-dc631cbc-6239-4cd6-95fd-9af3ead67d38.png)

kolibril13 commented:
Thanks for the review, @tylerjereddy .
I've now changed the logic in the voxelarray blog a bit, and I now use a for loop to assign values to the masked cube areas.
Further, I've changed the initial camera angle, increased the figure dpi to 250 and applied the black+flake8 linting issues.
Also, I've moved the example in the tutorial to the right place, in the first place, it was at gaussian_filter1d, but now it is at gaussian_filter
![image](https://user-images.githubusercontent.com/44469195/141744601-9aadd779-340a-43c0-92ab-23487eb62152.png)

kolibril13 commented:
It looks like the resolution did not change:
https://38675-1460385-gh.circle-artifacts.com/0/html-scipyorg/tutorial/ndimage.html#smoothing-filters
Maybe there is another option to show these plots a bit bigger?

tupui commented on doc/source/tutorial/examples/gaussian_filter_plot1.py:
```diff
@@ -0,0 +1,52 @@
+
+
+plt.rcParams["figure.dpi"] = 250
+fig = plt.figure(figsize=plt.figaspect(0.5))
+plot_voxels(voxelarray, fig, subp=(1, 3, 1), title="Original")
```
It would be better to use the OOP style a bit more by using `subplots` and pass axis to your plot function. There you can also set the aspect of the figure.
```suggestion
fig, axs = plt.subplots(1, 3, figsize=(12, 4))
plot_voxels(voxelarray, fig, subp= axs[0], title="Original")
```

tupui reviewed (commented):
Thanks @kolibril13 for the PR.

> It looks like the resolution did not change:
> https://38675-1460385-gh.circle-artifacts.com/0/html-scipyorg/tutorial/ndimage.html#smoothing-filters
> Maybe there is another option to show these plots a bit bigger?

You should be able to adjust the figure size using with something like `figsize=(12, 4)`. There should be no need to change the resolution.

kolibril13 commented on doc/source/tutorial/examples/gaussian_filter_plot1.py:
```diff
@@ -0,0 +1,54 @@
+    ax.set_title(title)
+
+
+fig = plt.figure(figsize=(12, 4))
```
@tupui : I think this is now the best solution

kolibril13 commented:
I locally run black and flake8 to forma the code, but it seems I am using another version of black than this repo, as there still was a white space linting error.
Because of that a suggestion: There is a github action https://pre-commit.ci/ that automatically fixes white space errors and applies code changes directly as a commit on a pull request branch, see e.g. here https://github.com/ManimCommunity/manim/pull/2280/commits/cc5b098fdb2096361cb8e18c52c4a4e3ac3a97e2

tupui commented:
> I locally run black and flake8 to forma the code, but it seems I am using another version of black than this repo, as there still was a white space linting error.
> 
> Because of that a suggestion: There is a github action https://pre-commit.ci/ that automatically fixes white space errors and applies code changes directly as a commit on a pull request branch, see e.g. here https://github.com/ManimCommunity/manim/pull/2280/commits/cc5b098fdb2096361cb8e18c52c4a4e3ac3a97e2

For various reasons (improper maths formatting being one), we don't want to use Black and pre-commit hooks. So any linter issues has to be resolved manually. Still some maintainers use Black and other tools and adjust the output to the desired state before pushing.

tupui commented on doc/source/tutorial/examples/gaussian_filter_plot1.py:
```diff
@@ -0,0 +1,43 @@
+voxelarray2 = gaussian_filter(voxelarray, sigma=1)
+plot_voxels(voxelarray2, ax2, title="gaussian_filter \n sigma=1")
+voxelarray3 = gaussian_filter(voxelarray, sigma=3)
+plot_voxels(voxelarray3, ax3, title="gaussian_filter \n sigma=3")
```
We also want some for of `tight_layout` I guess in order to not have the large padding. Also `show` so that we can run the script by hand and get the figure.
```suggestion
plot_voxels(voxelarray3, ax3, title="gaussian_filter \n sigma=3")

plt.tight_layout()
plt.show()
```

kolibril13 commented:
How it looks now:
https://38684-1460385-gh.circle-artifacts.com/0/html-scipyorg/tutorial/ndimage.html#smoothing-filters

tupui reviewed (commented):
The figure looks good now 😃 . If we want to be nitpicky, titles are a bit small.

The only thing left for me is to describe a bit more this figure for it to be useful. Right now it just says: `Here is an example`. Can you add 1-2 sentences about this case and the options that are used? In the description we talk about `order` and `sigma`. Depending on how you read this, a user could think that `order` was also changed.

kolibril13 commented:
Increase font size + more description

kolibril13 commented:
And here it is:
https://38686-1460385-gh.circle-artifacts.com/0/html-scipyorg/tutorial/ndimage.html#smoothing-filters

I was actually thinking, can we reuse this example to also explain the order parameter?
I've tried to apply it here:
```py
import matplotlib.pyplot as plt
import numpy as np
from scipy.ndimage import gaussian_filter

grids = 2
boxs = 5

voxelarray = np.zeros((boxs * grids, boxs * grids, boxs * grids))

i = 1
for xi in range(0, 2):
    for yi in range(0, 2):
        for zi in range(0, 2):
            voxelarray[
                xi * boxs: xi * boxs + boxs,
                yi * boxs: yi * boxs + boxs,
                zi * boxs: zi * boxs + boxs,
            ] = i
            i += 1

voxelarray = np.float64(voxelarray * 256 / 8)

with np.printoptions(threshold=4):
    print(voxelarray)
voxelarray=gaussian_filter(voxelarray, sigma=5, order=1)
with np.printoptions(threshold=4):
    print(voxelarray)
  ```
And here is the output. From my intuition, when having the derivative of the Gaussian, not every value should be zero. but it is. Do you have an idea why?
```
[[[ 32.  32.  32. ...  64.  64.  64.]
  [ 32.  32.  32. ...  64.  64.  64.]
  [ 32.  32.  32. ...  64.  64.  64.]
  ...
  [ 96.  96.  96. ... 128. 128. 128.]
  [ 96.  96.  96. ... 128. 128. 128.]
  [ 96.  96.  96. ... 128. 128. 128.]]

 [[ 32.  32.  32. ...  64.  64.  64.]
  [ 32.  32.  32. ...  64.  64.  64.]
  [ 32.  32.  32. ...  64.  64.  64.]
  ...
  [ 96.  96.  96. ... 128. 128. 128.]
  [ 96.  96.  96. ... 128. 128. 128.]
  [ 96.  96.  96. ... 128. 128. 128.]]

 [[ 32.  32.  32. ...  64.  64.  64.]
  [ 32.  32.  32. ...  64.  64.  64.]
  [ 32.  32.  32. ...  64.  64.  64.]
  ...
  [ 96.  96.  96. ... 128. 128. 128.]
  [ 96.  96.  96. ... 128. 128. 128.]
  [ 96.  96.  96. ... 128. 128. 128.]]

 ...

 [[160. 160. 160. ... 192. 192. 192.]
  [160. 160. 160. ... 192. 192. 192.]
  [160. 160. 160. ... 192. 192. 192.]
  ...
  [224. 224. 224. ... 256. 256. 256.]
  [224. 224. 224. ... 256. 256. 256.]
  [224. 224. 224. ... 256. 256. 256.]]

 [[160. 160. 160. ... 192. 192. 192.]
  [160. 160. 160. ... 192. 192. 192.]
  [160. 160. 160. ... 192. 192. 192.]
  ...
  [224. 224. 224. ... 256. 256. 256.]
  [224. 224. 224. ... 256. 256. 256.]
  [224. 224. 224. ... 256. 256. 256.]]

 [[160. 160. 160. ... 192. 192. 192.]
  [160. 160. 160. ... 192. 192. 192.]
  [160. 160. 160. ... 192. 192. 192.]
  ...
  [224. 224. 224. ... 256. 256. 256.]
  [224. 224. 224. ... 256. 256. 256.]
  [224. 224. 224. ... 256. 256. 256.]]]
[[[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 ...

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]

 [[0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  ...
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]
  [0. 0. 0. ... 0. 0. 0.]]]
  ```

tupui commented on doc/source/tutorial/ndimage.rst:
```diff
@@ -227,7 +229,13 @@ Smoothing filters
   derivatives are not implemented. The *order* parameter must be a
   number, to specify the same order for all axes, or a sequence of
-  numbers to specify a different order for each axis.
+  numbers to specify a different order for each axis. The below example 
+  shows the filter applied on test data with different sigma values. 
+  The order parameter is kept at 0. 
```
```suggestion
  numbers to specify a different order for each axis. The example below
  shows the filter applied on test data with different values of *sigma*.
  The *order* parameter is kept at 0. 
```

tupui reviewed (commented):
Some minor things I will commit for conveniency. For the order I cannot say. I am not using this and these is not much test I can rely on here. IMO this can be left for later as the current PR is already an improvement of the doc. @kolibril13 let me know if you want to investigate or if we can merge as it is.

kolibril13 commented:
It can be merged like this!
Thanks for your final improvements.

tupui commented:
Thanks @kolibril13 for the PR and Tyler for the review.
